backend/src/routes/blog.ts

```diff
@@ -320,4 +321,46 @@ blogRouter.get("/bulkUser/:id", async (c) => {
       message: "Error while fetching post",
     });
   }
+});
+interface ChatCompletionMessageParam {
+  role: 'system' | 'user' | 'assistant';
+  content: string;
+  name?: string;
+}
+blogRouter.post("/chat", async (c) => {
+  try {
+    if (!c.env.OPENAI_API_KEY) {
+      return c.json({
+        message: "This feature is disabled.",
+      });
+    }
+
+    const body = await c.req.json();
+    const { blogContent, userQuery, chatHistory, blogTitle } = body;
+
+    const openai = new OpenAI({
+      apiKey: c.env.OPENAI_API_KEY,
+    });
+
+    const messages: ChatCompletionMessageParam[] = [
+      { role: "system", content: "You are a helpful assistant that answers questions about blog content." },
+      {
+        role: "user",
+        content: `Title: ${blogTitle}\n\nBlog Content: "${blogContent}"\n\nPrevious Chat History:\n${chatHistory.map((message: ChatCompletionMessageParam) => `${message.role.toUpperCase()}: ${message.content}`).join("\n")}\n\nCurrent Question: ${userQuery}`,
+        name: "User"
+      },
+    ];
+
+    const response = await openai.chat.completions.create({
+      model: "gpt-3.5-turbo",
+      messages,
+    });
```

Comment:
For the purpose of code maintainability, let us separate out the openai GPT logic from the API route. 

There is a src/genAI.ts file within which you can keep the AI logic.

Reply:
Are we also ensuring the assistant does not answer questions that aren't related to the context?

If not, this can be done by adding a negative prompt within the the initial prompt.

Reply:
Separation of the AI logic is implemented. Also added some more information to prompt so that bot will not answer to anything if it is not related to blog content.
![Screenshot 2024-06-17 081801](https://github.com/aadeshkulkarni/figuringout/assets/107796595/9510f5c2-153d-4338-9e2c-006df31af591)